### DISCUSSION TITLE
MS SQL Server connector not working correctly with MSSQL v2012

### DISCUSSION THREAD

**Discussion by jeffprandall** [Budibase/budibase#3883] [Help] [answered]
I have started piloting Budibase and was going through the "Build a CRUD app - SQL" against a Microsoft SQL database and get to the point where after I add the autogenerated Screens my main Vehicles Table is blank.  It says No rows found.  If I press Create new I can add to the database successfully.  Interestingly thought when I add the Service Logs screens all the data in the Table appears.  

So then I pointed to an existing table in that same database and it also generated no rows in the Screen Table.

**Comment by mjashanks** [comment]
Hmm, that's strange. Can you paste a `CREATE TABLE...` script for one of the tables that does not return rows?

**Reply by jeffprandall** [reply]
```
CREATE TABLE [dbo].[Vehicles](
	[id] [int] IDENTITY(1,1) NOT NULL,
	[Registration] [text] NULL,
	[Make] [text] NULL,
	[Model] [text] NULL,
	[Colour] [text] NULL,
	[Year] [smallint] NULL,
PRIMARY KEY CLUSTERED 
(
	[id] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY]
) ON [PRIMARY] TEXTIMAGE_ON [PRIMARY]```

**Reply by jeffprandall** [reply]
It's also SQL Standard 2012 it that helps to know

**Reply by royaljeffr** [reply]
Also getting this error:
```
 ERROR [1640887454758] (30 on b3d416c9bd8e): RequestError: Invalid usage of the option next in the FETCH statement.
bbapps                |     Error: RequestError: Invalid usage of the option next in the FETCH statement.
bbapps                |         at /app/dist/integrations/microsoftSqlServer.js:85:23
bbapps                |         at Generator.throw (<anonymous>)
bbapps                |         at rejected (/app/dist/integrations/microsoftSqlServer.js:6:65)
bbapps                |         at runMicrotasks (<anonymous>)
bbapps                |         at processTicksAndRejections (internal/process/task_queues.js:95:5)
bbapps                | Trace: Error: RequestError: Invalid usage of the option next in the FETCH statement.
bbapps                |     at /app/dist/integrations/microsoftSqlServer.js:85:23
bbapps                |     at Generator.throw (<anonymous>)
bbapps                |     at rejected (/app/dist/integrations/microsoftSqlServer.js:6:65)
bbapps                |     at runMicrotasks (<anonymous>)
bbapps                |     at processTicksAndRejections (internal/process/task_queues.js:95:5) {
bbapps                |   expose: true,
bbapps                |   statusCode: 400,
bbapps                |   status: 400
```

**Reply by mjashanks** [reply]
Hi all.

Just as an update, we use a library called Knex to manage our SQL queries. 

The issue that you are having is described here: https://github.com/knex/knex/issues/1527 - it looks as though knex only supports very recent versions of SQL Server. 

It's a fantastic library by the way, not dunking on them!

We will need to look into a fix for this, and also come up with a list of SQL Server versions that we are willing to support. 

@royaljeffr what version of SQL Server are you using?

I will log an issue for this.

**Reply by royaljeffr** [reply]
We are running SQL2012 Standard SP4 version 11.0.7507.2 to be exact.  I have since spun up a Postgres database and started the "Build a CRUD app" and it's worked flawless.  Having a list of supported SQL would be nice.

**Reply by datdamnzotz** [reply]
Same error on 
Microsoft SQL Server 2016 (SP3) (KB5003279) - 13.0.6300.2 (X64) and
Microsoft SQL Server 2019 (RTM-CU15) (KB5008996) - 15.0.4198.2 (X64)

**Reply by datdamnzotz** [reply]
More info - looks like it is totally related to the order by bug in knex.
On Microsoft SQL Server 2019 (RTM-CU15) (KB5008996) - 15.0.4198.2 (X64)

Insert 5 records works fine:
```
DROP TABLE test
GO

CREATE TABLE [dbo].[Test](
	[id] [int] IDENTITY(1,1) NOT NULL,
	[Name] [nvarchar](max) NULL,
	[Number] [float] NULL,
	[Bit] [bit] NULL,
PRIMARY KEY CLUSTERED 
(
	[id] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
) ON [PRIMARY] TEXTIMAGE_ON [PRIMARY]
GO

INSERT INTO Test ([Name], [Number], [Bit])
VALUES ('test', '1', '1'),
('test', '2', '1'),
('test', '3', '1'),
('test', '4', '1'),
('test', '5', '1')
```
![image](https://user-images.githubusercontent.com/3913859/151856922-8e7fe88e-c14d-483f-8917-ff77d3869647.png)

Insert 30 and it fails.  Looks like it fails at 10 records.

```
INSERT INTO Test ([Name], [Number], [Bit])
VALUES ('test', '1', '1'),
('test', '2', '1'),
('test', '3', '1'),
('test', '4', '1'),
('test', '5', '1'),
('test', '6', '1'),
('test', '7', '1'),
('test', '8', '1'),
('test', '9', '1'),
('test', '10', '1'),
('test', '11', '1'),
('test', '12', '1'),
('test', '13', '1'),
('test', '14', '1'),
('test', '15', '1'),
('test', '16', '1'),
('test', '17', '1'),
('test', '18', '1'),
('test', '19', '1'),
('test', '20', '1'),
('test', '21', '1'),
('test', '22', '1'),
('test', '23', '1'),
('test', '24', '1'),
('test', '25', '1'),
('test', '26', '1'),
('test', '27', '1'),
('test', '28', '1'),
('test', '29', '1'),
('test', '30', '1')
```
![image](https://user-images.githubusercontent.com/3913859/151857264-d5cbffb5-5f8f-49a5-b46f-f1ae5fad97d6.png)
![image](https://user-images.githubusercontent.com/3913859/151857318-2440c016-220e-483f-b5f7-ae8d4b2fa1da.png)

Only other thing I can add is (and I am not sure of how this even works) Looks like budibase is using knex 0.95.6, and they have released 1.0.1 (which might have solved it?) Donno.
https://github.com/Budibase/budibase/blob/develop/packages/server/package.json#L101
https://github.com/knex/knex/releases/tag/1.0.1

Wish I knew a tad more about node app development so I could plop my own patch in.

**Reply by mike12345567** [reply]
@datdamnzotz just wanted to say thanks for all the debugging, helped me to find a solution which I've raised for PR!

**Reply by datdamnzotz** [reply]
@mike12345567 lemme know if you want to test prior versions,  I have a msdn license so I have access to all of them.

**Comment by datdamnzotz** [comment, marked as answer]
Thanks for the fix #4328  @mike12345567 : ] works wonderfully now.
